File: apc-pipeline/apc/core/task_spec.py
from __future__ import annotations
import re
from typing import Any, Literal
from pydantic import BaseModel, Field, model_validator


class OutputSpec(BaseModel):
    type: Literal["text", "json", "markdown", "table"] = "json"
    strict: bool = False
    schema_: dict[str, Any] | None = Field(default=None, alias="schema")
    model_config = {"populate_by_name": True}


class ReasoningSpec(BaseModel):
    required: bool = False
    visibility: Literal["visible", "hidden"] = "hidden"


class QualityWeights(BaseModel):
    accuracy: float = 0.5
    instruction_following: float = 0.2
    format: float = 0.15
    robustness: float = 0.1
    efficiency: float = 0.05


class CostLimits(BaseModel):
    max_input_tokens: int | None = None
    max_output_tokens: int | None = None
    max_latency_ms: int | None = None


class TaskSpec(BaseModel):
    task_id: str
    name: str
    version: str = "1.0.0"
    objective: str
    input: dict[str, Any] = Field(default_factory=dict)
    output: OutputSpec = Field(default_factory=OutputSpec)
    constraints: list[str] = Field(default_factory=list)
    reasoning: ReasoningSpec = Field(default_factory=ReasoningSpec)
    quality_weights: QualityWeights = Field(default_factory=QualityWeights)
    cost_limits: CostLimits = Field(default_factory=CostLimits)
    evaluation: dict[str, Any] = Field(default_factory=dict)
    examples: list[dict[str, Any]] = Field(default_factory=list)

    @model_validator(mode="after")
    def _validate(self) -> "TaskSpec":
        errors: list[str] = []
        if not self.task_id or not re.fullmatch(r"[a-z0-9_]+", self.task_id):
            errors.append(f"task_id 必须是非空的小写字母/数字/下划线，当前为 {self.task_id!r}")
        if not self.objective or not self.objective.strip():
            errors.append("objective 不能为空")
        if self.output.type == "json" and self.output.strict and not self.output.schema_:
            errors.append("严格 JSON 输出任务必须提供 output.schema（作为结构化输出的唯一来源）")
        total = self.quality_weights.accuracy + self.quality_weights.instruction_following + \
            self.quality_weights.format + self.quality_weights.robustness + self.quality_weights.efficiency
        if abs(total - 1.0) > 0.01:
            errors.append(f"quality_weights 权重之和必须为 1.0，当前为 {total:.4f}")
        for name in ("max_input_tokens", "max_output_tokens", "max_latency_ms"):
            v = getattr(self.cost_limits, name)
            if v is not None and v <= 0:
                errors.append(f"cost_limits.{name} 必须为正整数，当前为 {v}")
        if errors:
            raise ValueError("TaskSpec 校验失败：\n- " + "\n- ".join(errors))
        return self

    @classmethod
    def from_yaml(cls, path: str) -> "TaskSpec":
        import yaml
        with open(path, encoding="utf-8") as f:
            data = yaml.safe_load(f)
        try:
            return cls.model_validate(data)
        except ValueError as e:
            raise ValueError(f"{path}: {e}") from e

File: apc-service/src/main/java/com/apc/domain/TaskSpec.java
package com.apc.domain;

import com.fasterxml.jackson.annotation.JsonAlias;
import com.fasterxml.jackson.annotation.JsonProperty;
import jakarta.validation.constraints.NotBlank;
import java.util.*;

public class TaskSpec {
    @NotBlank private String taskId;
    @NotBlank private String name;
    private String version = "1.0.0";
    private String objective;
    private Map<String, Object> input = new HashMap<>();
    private OutputSpec output = new OutputSpec();
    private List<String> constraints = new ArrayList<>();
    private ReasoningSpec reasoning = new ReasoningSpec();
    private QualityWeights qualityWeights = new QualityWeights();
    private CostLimits costLimits = new CostLimits();
    private Map<String, Object> evaluation = new HashMap<>();

    public TaskSpec() {}
    public String getTaskId(){return taskId;} public void setTaskId(String v){taskId=v;}
    public String getName(){return name;} public void setName(String v){name=v;}
    public String getVersion(){return version;} public void setVersion(String v){version=v;}
    public String getObjective(){return objective;} public void setObjective(String v){objective=v;}
    public Map<String,Object> getInput(){return input;} public void setInput(Map<String,Object> v){input=v;}
    public OutputSpec getOutput(){return output;} public void setOutput(OutputSpec v){output=v;}
    public List<String> getConstraints(){return constraints;} public void setConstraints(List<String> v){constraints=v;}
    public ReasoningSpec getReasoning(){return reasoning;} public void setReasoning(ReasoningSpec v){reasoning=v;}
    public QualityWeights getQualityWeights(){return qualityWeights;} public void setQualityWeights(QualityWeights v){qualityWeights=v;}
    public CostLimits getCostLimits(){return costLimits;} public void setCostLimits(CostLimits v){costLimits=v;}
    public Map<String,Object> getEvaluation(){return evaluation;} public void setEvaluation(Map<String,Object> v){evaluation=v;}

    public static class OutputSpec {
        private String type = "json";
        private boolean strict;
        @JsonProperty("schema") @JsonAlias("schema_")
        private Map<String, Object> schema;
        public OutputSpec() {}
        public String getType(){return type;} public void setType(String v){type=v;}
        public boolean isStrict(){return strict;} public void setStrict(boolean v){strict=v;}
        public Map<String,Object> getSchema(){return schema;} public void setSchema(Map<String,Object> v){schema=v;}
    }
    public static class ReasoningSpec {
        private boolean required = false;
        private String visibility = "hidden";
        public ReasoningSpec() {}
        public boolean isRequired(){return required;} public void setRequired(boolean v){required=v;}
        public String getVisibility(){return visibility;} public void setVisibility(String v){visibility=v;}
    }
    public static class QualityWeights {
        private double accuracy = 0.5;
        private double instructionFollowing = 0.2;
        private double format = 0.15;
        private double robustness = 0.1;
        private double efficiency = 0.05;
        public QualityWeights() {}
        public double getAccuracy(){return accuracy;} public void setAccuracy(double v){accuracy=v;}
        public double getInstructionFollowing(){return instructionFollowing;} public void setInstructionFollowing(double v){instructionFollowing=v;}
        public double getFormat(){return format;} public void setFormat(double v){format=v;}
        public double getRobustness(){return robustness;} public void setRobustness(double v){robustness=v;}
        public double getEfficiency(){return efficiency;} public void setEfficiency(double v){efficiency=v;}
    }
    public static class CostLimits {
        private Integer maxInputTokens;
        private Integer maxOutputTokens;
        private Integer maxLatencyMs;
        public CostLimits() {}
        public Integer getMaxInputTokens(){return maxInputTokens;} public void setMaxInputTokens(Integer v){maxInputTokens=v;}
        public Integer getMaxOutputTokens(){return maxOutputTokens;} public void setMaxOutputTokens(Integer v){maxOutputTokens=v;}
        public Integer getMaxLatencyMs(){return maxLatencyMs;} public void setMaxLatencyMs(Integer v){maxLatencyMs=v;}
    }
}
